### ISSUE TITLE
Support schema bindings on Java 9

### ISSUE THREAD

**Issue by nicolaiparlog** [highsource/maven-jaxb2-plugin#120]
When running on Java 9, schema bindings are apparently not applied - instead the XSD's namespace is used to determine the package.

## Description

It makes no difference whether this block is present in the XSD...

```xml
<xsd:annotation>
	<xsd:appinfo>
		<jaxb:schemaBindings>
			<jaxb:package name="com.pany.pack.age" />
		</jaxb:schemaBindings>
	</xsd:appinfo>
</xsd:annotation>
```

... or this block in a `bindings.xjb` ...

```xml
<jaxb:bindings schemaLocation="that/other/schema.xsd" node="/xsd:schema">
	<jaxb:schemaBindings>
		<jaxb:package name="com.pany.pack.age" />
	</jaxb:schemaBindings>
</jaxb:bindings>
```

... in both cases, the binding is ignored.

## Experiments

I observed that the `xjc` command line tool in Java 8 and 9 exhibits the expected behavior and that the difference between the Maven run on Java 8 and 9 was that the former uses `rt.jar` as JAXB API, while the latter uses the plugin's dependency `javax.xml.bind:jaxb-api`.

It [took me a while](http://stackoverflow.com/q/43630262/2525313) but eventually I got the Java 9 Maven build to say _Java JAXB API is loaded from the [jrt:/java.xml.bind]_, i.e. it used the Java 9 module _java.xml.bind_ but that still didn't do the job.

## Sources

Maybe one of these can help (I'm shooting in the dark here, could be utterly irrelevant):

* [JEP 255](http://openjdk.java.net/jeps/255) (merges Selected Xerces 2.11.0 Updates into JAXP)
* [JEP 268](http://openjdk.java.net/jeps/268) (develops a standard XML Catalog API that supports the OASIS XML Catalogs standard, v1.1)
* Apache Camel had [another issue](https://issues.apache.org/jira/browse/CAMEL-10922) and ["resolved"](https://fisheye.apache.org/changelog/camel-git?cs=2468ba87b3f86ebec2d663ebf75a2b215e516700) it by using `org.apache.cxf:cxf-xjc-plugin` instead - maybe they're doing something differently?

Related to #114.

**Comment by highsource** [comment]
Java 9 is generally not supported yet.

**Comment by highsource** [comment]
See also #114.

**Comment by nicolaiparlog** [comment]
The question is whether that will change. :smiley:

**Comment by aggregat4** [comment]
@highsource Hi Lexi! Boris from Disy here. Do you already have a plan for Java 9 compatibility? Is there any way we can support work on this? (Bounty, Contract work, ...) Or if you have any pointers to potentially necessary implementation changes we could evaluate whether we can invest some time into it.

**Comment by highsource** [comment]
@aggregat4 Hi Boris. :)

I don't have any plans for Java 9 yet, so I can't say anything specific. :(

I think the best would be if some of your devs could check what the specific problems with Java 9 are. We could then think of how to solve them.

**Comment by nicolaiparlog** [comment]
I looked at this long and hard together with @aggregat4 but we did not make any meaningful progress on the plugin. What we found out:

* the `xjc` command line tool works on Java 8 and Java 9 as expected
* the API used by the plugin (`org.glassfish.jaxb:jaxb-xjc:2.2.11`) works as expected if updated to Java 9 compatible version `org.glassfish.jaxb:jaxb-xjc-jdk9:2.3.0-b170127.1453`
* using that version with the plugin does _not_ work

As it stands I assume I am not using the API in the exact same way as the plugin does (although I tried to copy it). Maybe you can have a look? These two projects might help:

* [Java 9 WTF / Maven JAXB 2 Plugin](https://github.com/CodeFX-org/java-9-wtf/tree/master/maven-jaxb2-plugin): shows that the bare API works but the plugin doesn't; includes documentation on how to run it on Java 8 and 9
* [Maven JAXB2 Plugin / Java 9 Branch](https://github.com/nicolaiparlog/maven-jaxb2-plugin/tree/java-9): I forked and branched your project to create a test failing on Java 9 and also update some dependencies in the hope to get it to work

**Comment by aggregat4** [comment]
@highsource our current state of investigation on this issues is the following:

- As noted by @nicolaiparlog above we have created a testcase to demonstrate that the issue exists for the plugin
- We were not able to reproduce the issue by making a small test program that uses the XJC API in a way that is similar to how the jaxb2 plugin does it
- We debugged the jaxb2 pluging trying to identify a potential cause for the difference in behaviour, but were sadly unsuccessful

Our question now is: Could we count on some support from your side to assist in identifying the problem? Would there be any way to motivate you to do this (monetary? cake? ;) )? If so, when would you have time?

If you don't see a possibility there, we will probably write a minimal maven plugin ourselves that just shells out to the XJC command line tool.

**Comment by highsource** [comment]
This will be solved (no additional motivation required), but I can't promise a specific date.
I'll prioritize it over other tasks.

To be sure - the specific problem you have is that bindings not applied, correct?

**Comment by nicolaiparlog** [comment]
That's great news! :smiley: 

> To be sure - the specific problem you have is that bindings not applied, correct?

Yes. I just opened #126 to share what I've got so far, including a test case showing that class and package names are ignored. The binding generating `is...Set` methods is ignored as well (not visible in that PR).

**Comment by highsource** [comment]
I am positively puzzled. Bindings seem to be processed (wrong element names produce an error) but not applied.

**Comment by highsource** [comment]
For some reason when schema is loaded, `jaxb:schemaBindings` is not parsed to `BISchemaBinding`. Internally, schema loaded uses JAXB for parsing. so `jaxb:schemaBindings` should be unmarshalled to `BISchemaBinding`, but this is not happening.

Digging further I've found that there are differences in `rootMap` of the JAXB context. Normally it should be:
```
{["http://java.sun.com/xml/ns/jaxb","factoryMethod"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@1e8ce150],["http://java.sun.com/xml/ns/jaxb","inlineBinaryData"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@604f2bd2],["http://java.sun.com/xml/ns/jaxb","javaType"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@1d3ac898],["http://java.sun.com/xml/ns/jaxb/xjc","javaType"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@1b73be9f],["http://java.sun.com/xml/ns/jaxb","class"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@628c4ac0],["http://java.sun.com/xml/ns/jaxb","property"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@7b84fcf8],["http://java.sun.com/xml/ns/jaxb/xjc","substitutable"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@30b19518],["http://java.sun.com/xml/ns/jaxb","globalBindings"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@363042d7],["http://java.sun.com/xml/ns/jaxb","typesafeEnumMember"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@366ac49b],["http://java.sun.com/xml/ns/jaxb","schemaBindings"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@6ad59d92],["http://www.w3.org/2001/XMLSchema","annotation"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@56f0cc85],["http://java.sun.com/xml/ns/jaxb","dom"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@62e20a76],["http://java.sun.com/xml/ns/jaxb/xjc","dom"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@2cc44ad],["http://java.sun.com/xml/ns/jaxb","typesafeEnumClass"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@44b3606b]}
```

But it is:

```
{["","factoryMethod"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@5f7eee96],["","inlineBinaryData"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@3a36cd5],["","javaType"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@53f0d09c],["http://java.sun.com/xml/ns/jaxb/xjc","javaType"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@47acd13b],["","class"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@6f8e9d06],["","property"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@77d381e6],["http://java.sun.com/xml/ns/jaxb/xjc","substitutable"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@2272cbb0],["","globalBindings"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@3f6f3cc],["","typesafeEnumMember"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@180b3819],["","schemaBindings"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@733c464f],["http://www.w3.org/2001/XMLSchema","annotation"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@47272cd3],["","dom"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@73fbdf68],["http://java.sun.com/xml/ns/jaxb/xjc","dom"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@32f1fafe],["","typesafeEnumClass"=com.sun.xml.bind.v2.runtime.ClassBeanInfoImpl@236eccd1]}
```

Note that `schemaBindings` does not have namespace. This is probably the reason why it is not parsed.

Now the question is, why namespaces are missing. Namespace is provided by `package-info.java`. Maybe it is missing somehow?

**Comment by highsource** [comment]
Indeed, 

```
XmlSchema xs = reader().getPackageAnnotation(XmlSchema.class,clazz,this);
```
Returns `null`.
Obviously the reader somehow does not return the `@XmlSchema` package annotation, probably can't read `package-info.class` for whatever reason.

**Comment by highsource** [comment]
@nicolaiparlog 
@aggregat4 

Seems like I've managed to fix it. I guess there was a classloader issue which prevented annotations from `com.sun.tools.xjc.reader.xmlschema.bindinfo.package-info` to be loaded properly.

I've added preloading of this packaged and it worked. Now I can build the `java-9` test project on JDK 9.

Please check if this works for you (build from the `java-9` branch).

Don't close yet, I want to check why `episode` fails. But not today.

**Comment by nicolaiparlog** [comment]
I tried your changes (more precisely 1948a6b) and it indeed passes the test I wrote. Using it on the actual project failed, though:

[ERROR] : Provider class com.sun.xml.bind.v2.ContextFactory could not be instantiated: javax.xml.bind.JAXBException: ClassCastException: attempting to cast jar:file:/home/parlog/.m2/repository/javax/xml/bind/jaxb-api/2.3.0-b170201.1204/jaxb-api-2.3.0-b170201.1204.jar!/javax/xml/bind/JAXBContext.class to jrt:/java.xml.bind/javax/xml/bind/JAXBContext.class.  Please make sure that you are specifying the proper ClassLoader.

I did not have time to look into it yet or to reproduce it with a simple test case. Do you have an idea what could be going on here?

**Comment by nicolaiparlog** [comment]
By and large it looks good, but the problem above (as well as another one) persist. I hope to have some time next week to create a minimal, self-contained demonstration.

**Comment by highsource** [comment]
would be great

**Comment by estronque** [comment]
I found a similar problem about the schema bindings are apparently not applied. If same xsd file has several point in the filename for example STRANGE_PREFIX.FILE_NAME.XSD the xjc work but not applied the globalBindings like choiceContentProperty. I tested with java8 and java7.

**Comment by nicolaiparlog** [comment]
To give some context, we have 95 Maven projects that use this plugin, 71 of which rely on bindings. Before I opened this issue we could build the 24 non-binding projects. With your changes, we could build 68 of the remaining 71. :+1:

## Class Cast Exception

One of the remaining three projects threw the error above:

```
[ERROR] : Provider class com.sun.xml.bind.v2.ContextFactory could not be instantiated:
javax.xml.bind.JAXBException: ClassCastException: attempting to cast
jar:file:/home/parlog/.m2/repository/javax/xml/bind/jaxb-api/2.3.0-b170201.1204/jaxb-api-2.3.0-b170201.1204.jar!/javax/xml/bind/JAXBContext.class
to jrt:/java.xml.bind/javax/xml/bind/JAXBContext.class.
Please make sure that you are specifying the proper ClassLoader.
```

It took me a while to figure out that the easiest fix is to no longer add the _java.xml.bind_ module to the Maven process. It's required with 0.13.2 but I assume the update to _jaxb-api-2.3.0-b170201.1204_ made that unnecessary. It's weird that the clash only happened in one project (it's the JAXB-heaviest, though) but I didn't investigate why.

## Stubborn Bindings

In the two remaining projects the plugin continued to not honor the bindings. We spent some looking at it but eventually pulled the plug and just made it work without them (it was just about a handful  `isSet...` methods). I did not generate a self-contained example, so I can't really provide further information.

There is [another project](https://www.disy.net/aktuelles/newsletter/newsletterartikel/artikel/2571.html) that is even heavier on JAXB, though, so I may have to face the music eventually. In that case, I'll be back with more information.

## What Now?

Bindings generally work now and I'm happy with the situation. :smiley: I think this issue can be closed once the solution is merged into master. I would file any further problems as a bug then.

**Comment by dzmitryh** [comment]
Hey folks, thanks at first for your efforts to make it complaint with java 9.
However even with plugin that built from java 9 branch(correct me if I'm wrong it haven't released yet) I'm getting this:

```
ERROR] Failed to execute goal org.jvnet.jaxb2.maven2:maven-jaxb2-plugin:0.13.3-SNAPSHOT:generate (default) on project some-project: Execution default of goal org.jvnet.jaxb2.maven2:maven-jaxb2-plugin:0.13.3-SNAPSHOT:generate failed: Prefix '' is already bound to '' -> [Help 1]
``` 
any clues what I can do with it ?

**Comment by nicolaiparlog** [comment]
That problem has something to do with episode files. I have no clue what those are and what they're used to, but the solution is to put `<epsiode>false</episode>` into the plugin configuration. (Sorry for not being more precise, I don't have it in front of me right now.)

**Comment by dzmitryh** [comment]
@nicolaiparlog thx for the tip. Applied your suggestion and it works! Thx a lot.

**Comment by dzmitryh** [comment]
@highsource Hi again. Just a quick question. Are there any plans to release java-9 branch of this plugin ? Thx in advance.

**Comment by highsource** [comment]
In principle, yes, I consider adding Java 9 support.

**Comment by paroxysm** [comment]
@nicolaiparlog @highsource We're using this plugin and we use episodes. Is there any fix that allows episodes for jdk 9?

**Comment by paroxysm** [comment]
I've built the java-9 branch and tried to generate. Still an error
"Execution generate of goal org.jvnet.jaxb2.maven2:maven-jaxb2-plugin:0.13.3-SNAPSHOT:generate failed: Prefix '' is already bound to '' -> [Help 1]"

FYI, running "maven-jaxb2-plugin-sample-episode-0.13.3-maven-src.zip" in the releases section also fails with the same error.

I decided to use <episode>false</episode>. It manages to get past that error but reports a new error
"META-INF\sun-jaxb.episode] with if-exists="true" attributes. Transformation failed with an unexpected error. javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: Premature end of file."

I can get rid of this error using, <addIfExistsToEpisodeSchemaBindings>false</addIfExistsToEpisodeSchemaBindings> option.
And it's able to generate and also compile(bindings seem to work too)


I tried to replicate the above steps w/ the non java-9 branch, it generates but bindings aren't applied.

**Comment by paroxysm** [comment]
I've explored other plugins in the mean time. It appears jaxb-maven-plugin is incompatible with JDK9 and cxf plugin has config issues we cannot employ due to the scale of our schemas. We'll halt JDK9 migration until this plugin is fully working on JDK9, episodes and all.

**Comment by ptahchiev** [comment]
I tried building the java-9 branch. I had to run it with -Dmaven.test.skip=true because there were failing tests. Both with this branch and with the master I get the same error:

````
Caused by: java.lang.IllegalArgumentException: Prefix '' is already bound to ''
	at com.sun.xml.txw2.StartTag.addNamespaceDecl(StartTag.java:191)
	at com.sun.xml.txw2.ContainerElement._namespace(ContainerElement.java:333)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

````

**Comment by hendriks73** [comment]
I was able to get the `java-9` branch to work under Java9, by explicitly setting:

    <episode>false</episode>

in the plugin's `<configuration>` element. This was mentioned before in https://github.com/highsource/maven-jaxb2-plugin/issues/120#issuecomment-366549503. Apparently the [episode emitter](https://github.com/javaee/jaxb-v2/blob/master/jaxb-ri/xjc/src/main/java/com/sun/tools/xjc/addon/episode/PluginImpl.java) is broken or is given faulty data.

Here's a more complete excerpt from my `pom.xml`:

    <build>
        <plugins>
            <plugin>
                <groupId>org.jvnet.jaxb2.maven2</groupId>
                <artifactId>maven-jaxb2-plugin</artifactId>
                <configuration>
                    <schemaDirectory>schema_dir</schemaDirectory>
                    <schemaIncludes>
                        <include>my_schema.xsd</include>
                    </schemaIncludes>
                    <generatePackage>com.mypackage</generatePackage>
                    <markGenerated>true</markGenerated>
                    
                    <episode>false</episode>
                </configuration>
                [...]
            </plugin>
        </plugins>
    </build>


You also need to introduce these *explicit* dependencies:

            <dependency>
                <groupId>javax.xml.bind</groupId>
                <artifactId>jaxb-api</artifactId>
                <version>2.3.0</version>
            </dependency>
            <dependency>
                <groupId>org.glassfish.jaxb</groupId>
                <artifactId>jaxb-runtime</artifactId>
                <version>2.3.0</version>
            </dependency>
            <dependency>
                <groupId>javax.activation</groupId>
                <artifactId>javax.activation-api</artifactId>
                <version>1.2.0</version>
            </dependency>

Note that these dependencies are already compiled with Java9, i.e. they cannot be used under Java8, because of incompatible class version numbers.

**Comment by highsource** [comment]
I think an own "flavor" of maven-jaxb2-plugin will be neede for Java 9.

**Comment by karteeksaragad** [comment]
If we put <episode>false</episode> then it is not possible to generate the sources in different packages.

In our project, we have a requirement like... a schema (a.xsd) is importing another schema (b.xsd) and we have a binding file (.xjb) where multiple schema bindings are defined to generate each of the a.xsd and b.xsd schemas in different custom packages...it was all working fine in java 8 using maven-jaxb2 plugin but after upgrading to java 9 we have issues with this episodic compilation ......  We have parked our java 9 upgrade plans aside until we hear from nichol or somebody from maven-jaxb2 team to support this feature again.

**Comment by Nasminack** [comment]
Facing similar issue in our project after upgrading to java 9. 
maven-jaxb2 plugin works well with java8 but with java9 episodic compilation is not working.  

"prefix ' 'already bound to exception" occuring in maven - jaxb2 plugin +java9 (eventhough <episode>false</episode> fix this,but it disables episodic compilation )

We require Episodic compilation as we have multiple schemas defined in the binding file and require java classes to be generated in different custom packages.

Please let us know when  maven - jaxb2 plugin with  java9 will be released!!!!!

**Comment by highsource** [comment]
@nicolaiparlog @dzmitryh @hendriks73 @ptahchiev 

Hi everyone,

I've just added the `maven-jaxb23-plugin` on the `java-9` branch. With this plugin, the `tests/java-9` by @nicolaiparlog more or less works. You still have to turn off episode generation, but schema bindings are considered.

Could you please give it a try on your projects. Thank you.

Best wishes,
Alexey

**Comment by highsource** [comment]
@nicolaiparlog @paroxysm @hendriks73 @karteeksaragad @Nasminack 

Please see #146 for the episodes problem. And I think I have solved it in the meanwhile.

A strange effect I see is that XJC somehow fails reading annotations form `package-info` correctly. The consequence is that XJC associates wrong namespace with the package and this breaks things.

However, if *I* try to read the same annotation in *my* code, I get correct results - and then, completely surprizing, XJC code works as well. This is above my understanding, why. Probably there's some caching going on so loading annotations in my code "warms up" it all somehow. I totally puzzled. I don't even know how to create a MCVE to ask on Stackoverflow.

Anyway, please try the current `java-9` branch (use `maven-jaxb23-plugin`). `tests/java-9` works with it completely (with turned on episode file generation). Please let me know if you have any further issues.

**Comment by nicolaiparlog** [comment]
Great work Alexey, thank you! :+1:

**Comment by highsource** [comment]
@nicolaiparlog Works for you?

**Comment by karteeksaragad** [comment]
Good work Alexey!!  Finally you made our day. Thank you very much.  We are now able to successfully get rid of "prefix ' 'already bound to exception" with your maven-jaxb23-plugin's java9/0.13.3-SNAPSHOT code base.  "Episodic compilation with multiple dependent schemas defined in the binding file which require java classes to be generated in different custom packages" is now working with java 9.

Couple of questions below, please don't  mind answering....
1. When are you releasing this java9's maven-jaxb23-plugin officially (non snapshot release).
2. Is it possible for you to make this maven-jaxb23-plugin compatible with java8 and java9 using some maven profiles internally.

Thanks,
karteeksaragad / Nasminack.

**Comment by karteeksaragad** [comment]
1. I understand that, but any tentative date will help us plan our java9 migration activities.
2. I think we need to find a way to make our projects work between java8/java9 by a switching mechanism, because for java9 we have added --add-modules in maven-compiler/maven-surefire...etc to make our build work...which java8 don't understand.

**Comment by giacgbj** [comment]
JFYTK, please consider that Java 9 has been already phased out and Java 10 end of public updates will be this September 
(http://www.oracle.com/technetwork/java/eol-135779.html)

**Comment by karteeksaragad** [comment]
Aware of that.

**Comment by highsource** [comment]
@karteeksaragad 

1. OK. Let's say I'll wait for feedback for another week. If there will be no negaive feedback till then, I'll release on 12-13.05. There will be no special Java 9 version first, `maven-jaxb23-plugin` will use Java 9 dependencies of JAXB, `maven-jaxb2-plugin` will be a synonym to `maven-jaxb2-plugin`. This is how it will be released, then we'll see if someone cries and try to fix if it happens. Sorry, I don't see a much better way to ensure the whole thing works on the variety of JDKs and OSs out there.
2. Ahm, I don't see it in the scope of `maven-jaxb2-plugin` how you make your projects Java 8/Java9 compatible, switching mechanisms etc. At the moment `maven-jaxb23-plugin` seems to work fine both on Java 8 and Java 9. At least there were no reports that it does not so far. If you have to do something special for Java 9 - well, do it. Does it influence the `maven-jaxb2-plugin` somehow?

**Comment by nicolaiparlog** [comment]
I'll take it for a spin on Monday or Tuesday - promised. :)

**Comment by karteeksaragad** [comment]
1. Perfect. We will plan our migration tasks accordingly. "maven-jaxb2-plugin will be a synonym to maven-jaxb2-plugin"  is there a typo.  It should be "maven-jaxb2-plugin will be a synonym to maven-jaxb23-plugin" i guess. Correct me if i am wrong.

2. Yeah, maven-jaxb doesn't really depends on --add-modules. Mainly maven-compiler and maven-surefire plugins needs those --add-modules as argLine. We will see how to add a switching/profile mechanism for java8/java9 for those plug-ins.

**Comment by highsource** [comment]
@karteeksaragad 
1. Correct, a typo.
2. OK. Would be good if you report it then, I'll add this to sample projects.

**Comment by karteeksaragad** [comment]
Let me know if you have any contacts from maven-compiler & maven-surefire plugin developers, I will check what is the best way to add these "--add-module" dependencies to make our projects work with java8/java9...etc seamlessly.

**Comment by nicolaiparlog** [comment]
While running a build of our project with the newest snapshot (plugin 2.3 from java-9 branch) I reread this issue. As I see it these are three things I can provide feedback on.

## Do the stubborn bindings work now?

In [this comment](https://github.com/highsource/maven-jaxb2-plugin/issues/120#issuecomment-319393119) I described some bindings that refused to work even with your earlier fixes. I just checked again and they are indeed working properly now.

## Does the build work on Java 8 with the same plugin version Java 8?

Yes, I just ran a complete built in Java 8 with the snapshot version and observed no problems.

## Maven process arguments

While working on this I also discovered that the command line arguments that add modules to the Maven process are no longer needed. That is great news because on Java 11 these modules are gone.

Executing the plugin still leads to warnings about illegal accesses, though, and that will become a problem in a future Java version. I opened #148 to track that.


As I see it, this issue can be closed.

**Comment by karteeksaragad** [comment]
@nicolaiparlog,

Quick question on your   #3: Maven process arguments,

Do you mean that we don't need to pass any command line arguments (--add-modules) to this maven-jaxb23 plugin or in general any maven plugin.  In our case, we have issues with "maven-compiler/maven-surefire" plugins if we don't pass these --add-modules.

**Comment by ptahchiev** [comment]
I can't build the `java-9` branch. For me it fails with:
````
[INFO] --- maven-plugin-plugin:3.5:helpmojo (help-goal) @ maven-jaxb23-plugin ---
[ERROR] 

Artifact Ids of the format maven-___-plugin are reserved for 
plugins in the Group Id org.apache.maven.plugins
Please change your artifactId to the format ___-maven-plugin
In the future this error will break the build.


[INFO] Using 'UTF-8' encoding to read mojo source files.
[INFO] java-javadoc mojo extractor found 0 mojo descriptor.
[ERROR] Failed to analyze /home/petar/.m2/repository/org/glassfish/jaxb/jaxb-xjc-jdk9/2.3.0/jaxb-xjc-jdk9-2.3.0.jar!/com/sun/tools/xjc/XJC2Task.class
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Maven JAXB 2.x Plugin Project ...................... SUCCESS [  0.311 s]
[INFO] Maven JAXB 2.x Plugin Core ......................... SUCCESS [  1.450 s]
[INFO] Maven JAXB 2.0.x Plugin ............................ SUCCESS [  1.654 s]
[INFO] Maven JAXB 2.1.x Plugin ............................ SUCCESS [  0.779 s]
[INFO] Maven JAXB 2.2.x Plugin ............................ SUCCESS [  0.853 s]
[INFO] Maven JAXB 2.3.x Plugin ............................ FAILURE [  0.108 s]
[INFO] Maven JAXB 2.x Plugin .............................. SKIPPED
[INFO] Maven JAXB 2.x Plugin Testing ...................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.464 s
[INFO] Finished at: 2018-05-08T09:42:13+03:00
[INFO] Final Memory: 33M/611M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.5:helpmojo (help-goal) on project maven-jaxb23-plugin: Execution help-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.5:helpmojo failed.: IllegalArgumentException -> [Help 1]
````

**Comment by nicolaiparlog** [comment]
@karteeksaragad: I can only speak for our build, which was successful without using `--add-modules` **for the Maven process** (e.g. with `.mvn/jvm.config` or `MAVEN_OPTS`). If you still need it for Surefire, then it looks like your tests need a Java EE module - [replace it with a third-party dependency](https://stackoverflow.com/a/48204154/2525313).

@ptahchiev: I got an error (not sure whether it's the same one) when I tried building with Java 9. Building with Java 8 succeeded.

**Comment by ptahchiev** [comment]
@nicolaiparlog for me it was the opposite. The exception was thrown when using java-8. With java-9 I managed to build the plugin just fine, and also my project builds just fine. Thank you guys :+1:

**Comment by highsource** [comment]
@ptahchiev Please post the full `mvn clean install -e -X` log.

**Comment by highsource** [comment]
This is crazy.
I just can't get the build work under Java 9.

**Comment by highsource** [comment]
Forget the release.
Sorry.

**Comment by paroxysm** [comment]
What's going on? Is java 9 support working OR were there complications?

**Comment by highsource** [comment]
There were complications. I can't build the project under Java 9.

**Comment by nicolaiparlog** [comment]
Me neither, but that's not necessary for the release, is it?

**Comment by highsource** [comment]
Ok, sorry everybody I got a bit emotional.

For the release I have to make the build work under Java 9. The release should work in one batch process. This means I can't really switch the JDK version in the process. Since `maven-jaxb23-plugin` must be build under Java 9 this means the whole thing must be built under Java 9.

Unfortunately, Java 9 build does not work. I get different errors, mostly related to access to `javax.xml.bind` which I have no idea how to fix. I saw documentation [like this one](https://blog.codefx.org/tools/maven-on-java-9/), but it's just some wild juggling of command line arguments.

So unfortunately this is going to be a long tedious process of step-by-step fixes. I am sorry this won't just a quick release as I thought and I can't name the release date.

**Comment by buko** [comment]
Most errors relating to javax.xml.bind can be fixed by adding dependencies to the JAXB apis and implementation:


<dependency>
			<groupId>javax.xml.bind</groupId>
			<artifactId>jaxb-api</artifactId>
		</dependency>
		<dependency>
			<groupId>org.glassfish.jaxb</groupId>
			<artifactId>jaxb-runtime</artifactId>
		</dependency>

You’ll need to look up the versions correct for you.

Even the short command line hacks around –add-module are only temporary fixes.

**Comment by highsource** [comment]
@nicolaiparlog Thank you!

I will definitely be working on this. Just it probably won't be as fast as as I thought. Sorry.

**Comment by highsource** [comment]
@nicolaiparlog Thanks, I'm fine financially.

**Comment by highsource** [comment]
There's progress. Build under Java 9 slowly starts to work. Most of the project can be build under Java 9 now, around half of the test projects work. I'll continue working on it the next days.

**Comment by highsource** [comment]
@nicolaiparlog @ptahchiev @aznan2 Please try building the maste branch under Java 9:

    mvn clean install -Pall -DperformRelease

Works for me now.

**Comment by aznan2** [comment]
The build works using a normal `mvn clean install`, but fails when adding the `-Pall`-flag: https://gist.github.com/aznan2/290dc67df51893db21e06a1d9a2eee12

**Comment by highsource** [comment]
@aznan2 Hm, I wonder how comes that you [have `0.9.2-SNAPSHOT` as version in the failing test project](https://gist.github.com/aznan2/290dc67df51893db21e06a1d9a2eee12#file-mvn-log-L1772).

```
[ERROR] Error resolving dependency resource [Dependency {groupId=org.jvnet.jaxb2.maven2, artifactId=maven-jaxb2-plugin-tests-MAVEN_JAXB2_PLUGIN-53-a, version=0.9.2-SNAPSHOT, type=jar, classifier=null, resource=a.xsd}].
```

**Comment by aznan2** [comment]
@highsource It seems catalog file in tests/MAVEN_JAXB2_PLUGIN-53/b/src/main/resources/catalog.cat contains the line
    PUBLIC "http://maven-jaxb2-plugin/samples/episode/a" "maven:org.jvnet.jaxb2.maven2:maven-jaxb2-plugin-tests-MAVEN_JAXB2_PLUGIN-53-a:jar::0.9.2-SNAPSHOT!/a.xsd"
If I change 0.9.2 to 0.14.0 the build works.

**Comment by highsource** [comment]
@aznan2 Ah, right. Then it's clear why it's failing for you and not for me: I've got this snapshot version from the (very) old builds. I've corrected this, please try again.

**Comment by aznan2** [comment]
@highsource Yes, it works now. Woo!

**Comment by kevemueller** [comment]
Just compiled master inside Eclipse with JDK10 and am using it in my project. Both manual as well as automated (m2e) runs are working fine.  The bindings.jxb file is properly applied.
Thanks for all the hard work you have put into this project!

Initial issues with the automated run (m2e) I could fix by replacing
--add-modules=ALL-SYSTEM   (throws API incompatibility exception as rt loads bundled jaxb)
to 
--add-modules=java.xml.ws.annotation (otherwise Eclipse won't start)

I suggest you push a release, so you get a wider coverage of bugs.

**Comment by adlmez** [comment]
+1

**Comment by adlmez** [comment]
Thanks a lot !

**Comment by highsource** [comment]
@aggregat4 @paroxysm @adlmez @aznan2 @kevemueller 

I've just released `maven-jaxb2-plugin` version 0.14.0 to the Central Maven Repo. Will appear there in an hour or so.

Please test it. `maven-jaxb2-plugin` or `maven-jaxb23-plugin` should work under Java 9 as well as 1.7 and 1.8.

`maven-jaxb20-plugin`, `maven-jaxb22-plugin`, `maven-jaxb22-plugin` should work under 1.7 and 1.8.

Java 1.6 is no longer supported.

**Comment by paroxysm** [comment]
@highsource cheers, will test soon.

**Comment by divanmostert** [comment]
Hi Guys,

@highsource is multiple namespaces supported at the moment? Also, one importing another?

If so, do I need a special setting to enable it?

**Comment by divanmostert** [comment]
Hi Guys,

@highsource is multiple namespaces supported at the moment? Also, one importing another?

If so, do I need a special setting to enable it?

**Comment by highsource** [comment]
@divanmostert Multiple namespaces should be no problem. Do you experience any difficulties? Please open a separate issue if you do.

**Comment by divanmostert** [comment]
Ok, thanks - just scaling down my problem, so will be in touch if I can't resolve it.

**Comment by pnabbefeld** [comment]
Hello, just stumbling in because I've got an error message because of "com.sun.org.apache.xml.internal.resolver.tools.CatalogResolver" being unaccessible as a superclass of MavenCatalogResolver - the internal class is not exported within Java 9 by default, and it seems it's no more available in JDK10 (looked only at src.jar, as I don't really use it, yet).

For JDK 9, I added these parameters to .mvn/jvm.config:
--add-opens java.base/java.lang=ALL-UNNAMED
--add-exports java.xml/com.sun.org.apache.xml.internal.resolver.tools=ALL-UNNAMED
--add-exports java.xml/com.sun.org.apache.xml.internal.resolver=ALL-UNNAMED
--illegal-access=deny

Plugin dependency:
                    <dependency>
                        <groupId>org.glassfish.jaxb</groupId>
                        <artifactId>jaxb-xjc-jdk9</artifactId>
                        <version>2.3.0.1</version>
                    </dependency>

This combination works for me (I know, the dependency is mentioned in the thread before, but so it's easier to copy for others looking here for support  ;-)  )

**Comment by darmbrust** [comment]
Just in case anyone stumbles over this bug, while trying to figure out their `unexpected element (uri: `errors with JDK 9, 10 or 11, there is something broken in the jdk that is making it ignore the package-info.java file in some cases.  This question finally got me to a solution:

https://stackoverflow.com/questions/52157040/jaxb-package-info-ignored-when-using-java-10

and the magic solution, if you are using this plugin:

`<packageLevelAnnotations>false</packageLevelAnnotations>`

since there seems to be a nasty regression in the jdk, or something about how maven is setting up the classloader when I run my code inside a mojo.

For me, the kicker was I could unmarshal my XML using classes generated from schema by maven, if I ran my code within eclipse.  But, if I ran my code as a maven plugin, it failed.  The way they are searching for the package-info.java file doesn't work when the classes are loaded by the maven class loader.
